File: go_server/http_utils.go
package main

import (
	"bufio"
	"encoding/json"
	"net"
	"strconv"
	"strings"
)

type HttpMethod int8

const (
	Get HttpMethod = iota
	Put
	NotImplemented
)

type HttpRequest struct {
	Method  HttpMethod
	Path    string
	Headers []string
	Body    *Entity
}

func NewHttpRequest() *HttpRequest {
	return &HttpRequest{
		Method:  NotImplemented,
		Path:    "",
		Headers: make([]string, 0),
		Body:    nil,
	}
}

func ParseRequest(conn net.Conn) (*HttpRequest, error) {
	reader := bufio.NewReader(conn)

	request := NewHttpRequest()

	//Parse headers
	for {
		header, err := reader.ReadString('\n')
		if err != nil {
			return nil, ErrBadRequest{}
		}

		if header == "\r\n" {
			break
		}

		request.Headers = append(request.Headers, strings.TrimSpace(header))
	}

	//Parse request line
	request_line := request.Headers[0]
	request.Headers = request.Headers[1:]

	request_parts := strings.Split(request_line, " ")
	for idx, _ := range request_parts {
		request_parts[idx] = strings.TrimSpace(request_parts[idx])
	}

	switch request_parts[0] {
	case "GET":
		request.Method = Get
	case "PUT":
		request.Method = Put
	default:
		request.Method = NotImplemented
	}

	request.Path = request_parts[1]

	//Parse body
	contentLength := 0
	for _, header := range request.Headers {
		if strings.Contains(header, "Content-Length") {
			parts := strings.Split(header, ":")
			contentLengthStr := strings.TrimSpace(parts[1])
			var err error
			contentLength, err = strconv.Atoi(contentLengthStr)
			if err != nil {
				return nil, ErrBadRequest{}
			}
		}
	}

	if contentLength > 0 {
		bodyBytes := make([]byte, contentLength)
		readNum, err := reader.Read(bodyBytes)
		if err != nil || readNum != contentLength {
			return nil, ErrBadRequest{}
		}

		var entity Entity
		err = json.Unmarshal([]byte(bodyBytes), &entity)
		if err != nil {
			return nil, ErrJsonParsing{Message: err.Error()}
		}

		request.Body = &entity
	}

	return request, nil
}

File: rust_server/src/http_utils.rs
use std::io::{BufRead, BufReader, Read};
use std::net::TcpStream;
use crate::model::Entity;


#[derive(Debug)]
pub enum HttpMethod{
    Get,
    Put,
    NotImplemented
}
#[derive(Debug)]
pub struct HttpRequest{
    pub method: HttpMethod,
    pub path: String,
    pub headers: Vec<String>,
    pub body: Option<Entity>
}
impl HttpRequest {
    fn new() -> Self {
        Self{
            method: HttpMethod::NotImplemented,
            path: String::from(""),
            headers: vec![],
            body:None
        }
    }
}

pub fn parse_request(stream: &mut TcpStream) -> Result<HttpRequest, String>{
    let mut buff_reader = BufReader::new(stream);

    let mut request = HttpRequest::new();

    //Parsing header
    loop {
        let  line = &mut String::from("");
        if let Ok(_) = buff_reader.read_line(line){
            if line == "\r\n"{
                break;
            }
            request.headers.push(line.clone().trim().to_string());
        }
        else {
            let response = "HTTP/1.1 400 BadRequest \r\n\r\n".to_string();
            return Err(response);
        }
    }


    // Parsing request line
    if let Some(request_line) = &request.headers.first(){
        let parts: Vec<&str> = request_line.split(' ').collect();

        request.method =  match parts[0].trim()  {
            "GET" => HttpMethod::Get,
            "PUT" => HttpMethod::Put,
            _ => HttpMethod::NotImplemented
        };

        request.path = parts[1].trim().to_string();
    }
    else{
        let response = "HTTP/1.1 400 BadRequest \r\n\r\n".to_string();
        return Err(response);
    }
    request.headers.remove(0);




    //Parsing body
    //Must be done using content length because otherwise buffer cant find end of line for last line of body because
    //EOF is appended when connection is closed
    let mut content_length = 0;
    for line in &request.headers{
        if line.to_lowercase().contains("content-length"){
            let parts: Vec<&str> = line.split(":").collect();
            let len_str = parts[1].trim();
            content_length = len_str.parse::<i32>().unwrap();
        }
    }

    if content_length == 0 {
        return Ok(request);
    }

    let mut buffer =  vec![0;content_length as usize];
    let body_content;
    if let Ok(_) = buff_reader.read_exact(&mut buffer){
        body_content =String::from_utf8_lossy(&buffer).to_string();
    }
    else {
        let response = "HTTP/1.1 400 BadRequest \r\n\r\n".to_string();
        return Err(response);
    }

    //Parsing json body to domain model
    match  serde_json::from_str::<Entity>(&body_content){
        Ok(entity) =>{
            request.body = Some(entity);
        },
        Err(err) => {
            let response = format!("HTTP/1.1 400 BadRequest \r\n\r\nJson error: {err}");
            return Err(response);
        }
    };

    Ok(request)
}
